### PR TITLE
Guest Blog: Caching OpenAI Chat API Responses with LangChain and Xata - Rishi

### PR DESCRIPTION
**Pull request by alexfrancoeur**
## Summary

- Creating a branch from @rishi-raj-jain's original PR to generate a preview and modify if needed: https://github.com/xataio/mdx-blog/pull/216

### Blog Release Checklist

- \[ ] Release on: March 27, 2024
- \[ ] Published on dev.to (connected to Xata org)
- \[ ] OG image included
- \[ ] Social media is scheduled for this post

### CODE REVIEW

File: cache-openai-langchain.mdx [hunk 1]
```diff
@@ -0,0 +1,517 @@
...
+## Final Thoughts
+
+In this guide, you learned different ways of caching OpenAI API responses using LangChain callbacks and the caching layer. Using Xata, you’re able to interact with a Postgres serverless database to build a set of cache OpenAI API responses, and utilize Xata’s search capabilities to respond to user queries faster using those cache responses.
+
+For more detailed insights, explore the references cited in this post.
+
+| Resource                        | Link                                                          |
+| ------------------------------- | ------------------------------------------------------------- |
+| GitHub Repo                     | https://github.com/rishi-raj-jain/xata-cache-openai-langchain |
+| Caching - LangChain Chat Models | https://js.langchain.com/docs/modules/model\_io/chat/caching  |
```
Comment (rishi-raj-jain) [inline cache-openai-langchain.mdx:512]: can we update the link to https://js.langchain.com/docs/modules/model_io/chat/caching instead?

**Comment by rishi-raj-jain** [comment]
apart from that, LGTM

**Review by alexfrancoeur — COMMENTED** [review]
This looks great to me overall. I had one question around the diff formatting but other than that I think it's good to ship. If any of the @xataio/frontenders have a few spare minutes in the next few days, @CezzaineZ and I would appreciate a once-over. Otherwise, I say we're good to ship once we have an illustration for the blog

**Comment by alexfrancoeur** [comment]
Quick update -- we should have an illustration for tomorrow. I'll follow up again on the diff markdown

**Comment by alexfrancoeur** [comment]
@rishi-raj-jain I added an illustration and highlights for the diffs (true diff support coming soon). We'll get this merged shortly.

**Comment by alexfrancoeur** [comment]
Ah looks like there were a few more adjustments, let me change and I'll wait for a ✅ from you

**Comment by rishi-raj-jain** [comment]
@alexfrancoeur 

That's great.

Here's the only change that I think is pending apart from diff support.

Screenshot 2024-03-27 at 8 55 36 PM

File: cache-openai-langchain.mdx [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
```
Comment (alexfrancoeur) [inline cache-openai-langchain.mdx:512]:
```suggestion
| Caching - LangChain Chat Models | https://js.langchain.com/docs/modules/model_io/chat/caching  |
```

File: cache-openai-langchain.mdx [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+// ...
+```
+
+The code above, creates an endpoint `/query-callback-method` which can be `POST`-ed to. Then, it destructures `input` from the request body. Then, it sets the `Content-Type` and `Transfer-Encoding` header which indicate creation of a streaming response from this endpoint to the client. Further, it creates a LLM via a `ChatOpenAI` instance with LangChain using multiple callbacks as follows:
+
+- **handleLLMStart**: It’s called when the Chat Model begins working on a prompt
+- **handleLLMNewToken**: It’s called whenever the Chat Model in `streaming` mode generates a new token while processing a request
+- **handleLLMEnd**: It’s called when the Chat Model finishes processing a request to generate its final output
+
+Finally, it creates a single message conversation between user and AI, and awaits the complete response by AI to the user query.
```
Comment (miukimiu) [inline cache-openai-langchain.mdx:238]:
```suggestion
Finally, it creates a single message conversation between the user and AI, and awaits the complete response by AI to the user query.
```

**Review by rishi-raj-jain — APPROVED** [review]

File: cache-openai-langchain.mdx [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+  const chain = new ConversationChain({ llm: model });
+  await chain.call({ input });
+});
+
+// ...
+```
+
+The code above, starts with importing the `getXataClient` method from the generated `xata.js` file. Then, it creates a global instance of the Xata client. Further, it updates the LangChain callbacks as follows:
+
+- handleLLMStart: As the LLM is started to the respond to the user query, it first looks for a cached response in your Xata database’s `responses` table. This is implemented via the following:
```
Comment (miukimiu) [inline cache-openai-langchain.mdx:306]:
```suggestion
- handleLLMStart: As the LLM starts to respond to the user query, it first looks for a cached response in your Xata database’s `responses` table. This is implemented via the following:
```

**Review by miukimiu — APPROVED** [review]
LGTM! 🎉 

I just have two small suggestions.

**Comment by alexfrancoeur** [comment]
Social scheduled to go out in a few hours ✅ 

Thanks @rishi-raj-jain !

**Pull request merged**